File: api/src/Domain/Billing/Entities/Subscription.cs
using System;

namespace DetailingArsenal.Domain.Billing {
    public class Subscription : Entity<Subscription>, IBillingEntity {
        public string Status { get; set; }
        public Period TrialPeriod { get; set; }
        public Period Period { get; set; }
        public bool CancellingAtPeriodEnd { get; set; }
        public SubscriptionPlanReference PlanReference { get; }
        public BillingReference BillingReference { get; }

        public Subscription(string status, Period trialPeriod, Period period, bool cancellingAtPeriodEnd, SubscriptionPlanReference planReference, BillingReference billingReference) {
            Id = Guid.NewGuid();
            Status = status;
            TrialPeriod = trialPeriod;
            Period = period;
            CancellingAtPeriodEnd = cancellingAtPeriodEnd;
            PlanReference = planReference;
            BillingReference = billingReference;
        }

        public Subscription(Guid id, string status, Period trialPeriod, Period period, bool cancellingAtPeriodEnd, SubscriptionPlanReference planReference, BillingReference billingReference) {
            Id = id;
            Status = status;
            TrialPeriod = trialPeriod;
            Period = period;
            CancellingAtPeriodEnd = cancellingAtPeriodEnd;
            PlanReference = planReference;
            BillingReference = billingReference;
        }
    }
}

File: app/src/api/scheduling/billing/data-transfer-objects/subscription.ts
import { SubscriptionPlanPrice } from '@/api/scheduling/billing/data-transfer-objects/subscription-plan-price';
import { SubscriptionStatus } from '@/api/scheduling/billing/data-transfer-objects/subscription-status';
import moment from 'moment';
import { Period } from '@/api/scheduling/billing/data-transfer-objects/period';

/**
 * Subscription for a user. Associates a plan, and a specific price along with the current status of the plan.
 */
export class Subscription {
    constructor(
        public planName: string,
        public price: SubscriptionPlanPrice,
        public status: SubscriptionStatus,
        public trialPeriod: Period,
        public period: Period,
        public cancellingAtPeriodEnd: boolean
    ) {}

    get trialDaysRemaining() {
        const today = moment();
        const trialEnd = moment(this.trialPeriod.end);

        return trialEnd.diff(today, 'days');
    }
}
